### ISSUE TITLE
[Go] What is the best way to handle dictionary overflow?

### ISSUE THREAD

**Issue by lquerel** [apache/arrow#14670]
Arrow allows declaring a string dictionary with different indice type (e.g. uint8, uint16, ...). Unfortunately, sometimes we don't know in advance the exact cardinality of a column, so deciding in advance the proper indice type is not always feasible and we have to rely on some kind of adaptive approach. 

I'd like to automatically determine when a dictionary usage overflows (during insertion) and then fallback to a larger indice type or directly to a string when I've reached the maximum indice size I want for a specific application. So far, I haven't found a very effective way to detect this overflow. The `AppendString` method doesn't return an error in case of overflow. So I have to insert all my data, then check all my dictionaries and call the `Offset` method to check if the offset is greater than the maximum value for the current indice type. 

Is there a better approach to handling dictionary overflow with Go Arrow SDK?

**Comment by zeroshade** [comment]
@lquerel So the C++ implementation has an `AdaptiveInt` class that can be used for building a dictionary which is basically the approach you're suggesting, in that it swaps out for larger bitsize integers as the number of entries in the dictionary grows. This approach could certainly get implemented in the Go library too (and is likely the preferred solution). 

For now, you could check the max dict length for the size you use by checking `DictionaryType.IndexType.(arrow.FixedWidthDataType).BitWidth()` and simply comparing the length against `1 << bitWidth` or something like that. I would definitely say to add an explicit GitHub Issue for implementing the adaptive builder approach though as it's definitely a good idea to implement in the library.

**Comment by lquerel** [comment]
Thanks @zeroshade 

While waiting for better, I started working on something I called AdaptiveSchema which is basically wrapping a schema and exposing 3 methods.

```go
type AdaptiveSchema interface {
    Schema() *arrow.Schema
    DetectDictionaryOverflow(record arrow.Record) (overflowDetected bool, updates []*SchemaUpdate)
    UpdateSchema(updates []*SchemaUpdate) error
}
```

Each time a new `arrow.Record` is created, the `AdaptiveSchema` detects any dictionary overflow and returns true if at least one overflow has been detected as well as a list of objects representing the schema updates to be applied before replaying the creation of this record (`UpdateSchema` method). This is definitely not perfect but for now it should be enough. I can contribute put a link to this code here if you want.

Does `AdaptiveInt` fall back to a string data type when the dictionary size reaches a configurable size? In my context, having this capability is important because converting a very high cardinality column to a dictionary is inefficient.

**Comment by zeroshade** [comment]
I don't believe the current C++ `AdaptiveInt` would fallback when it reaches a configurable size. That said, the parquet library's dictionary encoders *do* fallback to a plain encoding if the dictionary reaches a configurable size. But that's only relevant if you're writing parquet files....

As far as your `AdaptiveSchema`, I think it might be better if the actual dictionary builders did the proper check for an overflow of dictionary size when new entries are added and somehow signal that to callers so that an overflow can be detected when you're building the dictionary rather than having to detect it afterwards. That said, being able to reproduce the creation of a record is a pretty cool idea.

**Comment by lquerel** [comment]
I totally agree with you about the right approach, but it would require adapting the existing builder code. This is something I might consider contributing at some point, but for the next 3-4 weeks I need to stay focused on the Open Telemetry side.
